cc_fuzz_test() breaks non-Clang builds

## Actual Behavior

The [docs for cc_fuzz_target](https://github.com/bazelbuild/rules_fuzzing#configuring-the-bazelrc-file) require you to add the following lines to your `.bazelrc`:

```
# Force the use of Clang for C++ builds.
build --action_env=CC=clang
build --action_env=CXX=clang++
```

This makes it impossible to build the repo with any compiler other than Clang. But I want my repo to be compatible with GCC, MSVC, and other compilers, for testing my code across compilers.

I tried specifying these actions for the `libfuzzer` configurations only:

```
# Force the use of Clang for fuzzer builds only.
build:asan-libfuzzer --action_env=CC=clang
build:asan-libfuzzer --action_env=CXX=clang++
```

However, I discovered that the `cc_fuzz_target()` would then fail to build with other compilers. For example, compiling with GCC I got the following error:

```
bazel-out/k8-fastbuild-ST-6de91758680c/bin/external/rules_fuzzing/fuzzing/replay/_objs/replay_main/replay_main.pic.o:replay_main.cc:function main: error: undefined reference to 'LLVMFuzzerInitialize'
```

This is unfortunate; even if nobody tries to build the fuzzing target directly, a broken target like this will make `bazel test ...` fail.

## Expected Behavior

Ideally, `cc_fuzz_target()` would automatically opt itself out of non-Clang builds using [target_compatible_with](https://docs.bazel.build/versions/main/be/common-definitions.html#common.target_compatible_with).

This would cause the fuzz targets to be automatically skipped by Bazel on non-Clang builds. Users could get the best of both worlds: the repo could continue to be compatible with any compiler, but Bazel would only attempt to build the fuzzing rules if the build is using Clang.

I tried doing this myself by adding `target_compatible_with` to my `cc_fuzz_target()` rule:

```
cc_fuzz_test(
    name = "file_descriptor_parsenew_fuzzer",
    srcs = ["file_descriptor_parsenew_fuzzer.cc"],
    deps = [
        "//:descriptor_upb_proto",
        "//:upb",
    ],
    target_compatible_with = [
        "@bazel_tools//tools/cpp:clang",
    ],
)
```

However this didn't work; Bazel seemed to skip the test no matter what I did. I tried both of these commands, and both of them skipped the fuzz test:

```
CC=clang CXX=clang++ bazel test ...
```

and

```
bazel test --compiler=clang ...
```

I'm probably doing something wrong at the Bazel level. If `cc_fuzz_target()` has this built in I wouldn't need to figure out the right incantation to get this working. :)

Technically, `clang` is only required for the libfuzzer configuration. So disabling the targets in non-clang configurations for all fuzzing engines seems a bit too broad.

The replay fuzzing engine should actually work with all major compilers, and can be used to turn all fuzz tests into regression tests.

But I see from @haberman's report that this doesn't happen:

> However, I discovered that the `cc_fuzz_target()` would then fail to build with other compilers. For example, compiling with GCC I got the following error:
> 
> ```
> bazel-out/k8-fastbuild-ST-6de91758680c/bin/external/rules_fuzzing/fuzzing/replay/_objs/replay_main/replay_main.pic.o:replay_main.cc:function main: error: undefined reference to 'LLVMFuzzerInitialize'
> ```

I think this part is problematic, since the symbol was supposed to be weakly defined: https://github.com/bazelbuild/rules_fuzzing/blob/b2b3efdcd70d319c6e0e9b687baff3b54e5a5711/fuzzing/replay/replay_main.cc#L25

Maybe #139 introduced an incompatibility with GCC? @fmeum would you like to look into this?

--- a/.github/workflows/bazel_test.yml
+++ b/.github/workflows/bazel_test.yml
@@ -126,3 +126,14 @@ jobs:
           bazel test --action_env=CC=gcc --action_env=CXX=g++ \
               --verbose_failures --test_output=all \
               //examples/...
+  regression_tests_mac:
+    name: Regression tests (macOS)
+    runs-on: macos-latest
+    timeout-minutes: 30
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Run regression tests on macOS
+        run: |
+          bazel test --verbose_failures --test_output=all \
+              -- //examples/... -//examples/java:all

--- a/.github/workflows/bazel_test.yml
+++ b/.github/workflows/bazel_test.yml
@@ -114,3 +114,15 @@ jobs:
           bazel test --verbose_failures --test_output=all \
               --build_tag_filters=fuzz-test --config=${{ matrix.config }} \
               //examples/java/...
+  regression_tests_gcc:
+    name: Regression tests (GCC)
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Run regression tests with GCC
+        run: |
+          bazel test --action_env=CC=gcc --action_env=CXX=g++ \
+              --verbose_failures --test_output=all \
+              //examples/...

--- a/BUILD
+++ b/BUILD
@@ -11,3 +11,8 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+config_setting(
+    name = "clang",
+    flag_values = {"@bazel_tools//tools/cpp:compiler": "clang"},
+)

--- a/examples/BUILD
+++ b/examples/BUILD
@@ -58,6 +58,10 @@ cc_fuzz_test(
 cc_fuzz_test(
     name = "fuzzed_data_provider_fuzz_test",
     srcs = ["fuzzed_data_provider_fuzz_test.cc"],
+    target_compatible_with = select({
+        "//:clang": [],
+        "//conditions:default": ["@platforms//:incompatible"],
+    }),
 )
 
 cc_fuzz_test(

--- a/fuzzing/engines/honggfuzz_launcher.sh
+++ b/fuzzing/engines/honggfuzz_launcher.sh
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-command_line="$(readlink -f ${HONGGFUZZ_PATH})"
+command_line="$(python3 -c 'import os, sys; print(os.path.realpath(sys.argv[1]))' ${HONGGFUZZ_PATH})"
 command_line+=("--workspace=${FUZZER_OUTPUT_ROOT}")
 
 if [[ -n "${FUZZER_SEED_CORPUS_DIR}" ]]; then

--- a/fuzzing/engines/jazzer_launcher.sh
+++ b/fuzzing/engines/jazzer_launcher.sh
@@ -16,7 +16,7 @@
 # engine. The launch configuration is supplied by the launcher script through
 # environment variables.
 
-command_line=("$(readlink -f ${FUZZER_BINARY})")
+command_line=("$(python3 -c 'import os, sys; print(os.path.realpath(sys.argv[1]))' ${FUZZER_BINARY})")
 
 # libFuzzer flags (compatible with Jazzer).
 

--- a/fuzzing/engines/libfuzzer_launcher.sh
+++ b/fuzzing/engines/libfuzzer_launcher.sh
@@ -16,7 +16,7 @@
 # libFuzzer engine. The launch configuration is supplied by the launcher
 # script through environment variables.
 
-command_line=("$(readlink -f ${FUZZER_BINARY})")
+command_line=("$(python3 -c 'import os, sys; print(os.path.realpath(sys.argv[1]))' ${FUZZER_BINARY})")
 
 # libFuzzer flags.
 

--- a/fuzzing/engines/replay_launcher.sh
+++ b/fuzzing/engines/replay_launcher.sh
@@ -16,7 +16,7 @@ if (( ! FUZZER_IS_REGRESSION )); then
     echo "NOTE: Non-regression mode is not supported by the replay engine."
 fi
 
-command_line=("$(readlink -f ${FUZZER_BINARY})")
+command_line=("$(python3 -c 'import os, sys; print(os.path.realpath(sys.argv[1]))' ${FUZZER_BINARY})")
 if [[ -n "${FUZZER_SEED_CORPUS_DIR}" ]]; then
     command_line+=("${FUZZER_SEED_CORPUS_DIR}")
 fi

--- a/fuzzing/private/fuzz_test.bzl
+++ b/fuzzing/private/fuzz_test.bzl
@@ -174,6 +174,7 @@ def cc_fuzz_test(
     # instrumented configuration.
     raw_binary_name = name + "_raw_"
     binary_kwargs.setdefault("deps", []).append(engine)
+    binary_kwargs.setdefault("tags", []).append("manual")
     cc_binary(
         name = raw_binary_name,
         **binary_kwargs
@@ -251,6 +252,7 @@ def java_fuzz_test(
         "Jazzer-Fuzz-Target-Class: %s" % target_class,
     ]
     binary_kwargs.setdefault("deps", []).append(engine)
+    binary_kwargs.setdefault("tags", []).append("manual")
     native.java_binary(
         name = raw_target_name,
         srcs = srcs,
@@ -282,6 +284,7 @@ def java_fuzz_test(
             "@rules_fuzzing//fuzzing/private:use_oss_fuzz": "@rules_fuzzing//fuzzing/private:oss_fuzz_jazzer_sanitizer_options.sh",
             "//conditions:default": "@rules_fuzzing//fuzzing/private:local_jazzer_sanitizer_options.sh",
         }),
+        tags = ["manual"],
         target = raw_target_name,
         target_deploy_jar = raw_target_name + "_deploy.jar",
     )

--- a/fuzzing/private/java_utils.bzl
+++ b/fuzzing/private/java_utils.bzl
@@ -110,7 +110,7 @@ runfiles_export_envvars
 # Determine the path to load libjvm.so from, either relative to the location of
 # the java binary or to $JAVA_HOME, if set. On OSS-Fuzz, the path is provided in
 # JVM_LD_LIBRARY_PATH.
-JAVA_BIN=$(readlink -f "$(which java)")
+JAVA_BIN=$(python3 -c 'import os, sys; print(os.path.realpath(sys.argv[1]))' "$(which java)")
 JAVA_HOME=${JAVA_HOME:-${JAVA_BIN%/bin/java}}
 # The location of libjvm.so relative to the JDK differs between JDK <= 8 and 9+.
 JVM_LD_LIBRARY_PATH=${JVM_LD_LIBRARY_PATH:-"$JAVA_HOME/lib/server:$JAVA_HOME/lib/amd64/server"}

--- a/fuzzing/replay/replay_main.cc
+++ b/fuzzing/replay/replay_main.cc
@@ -22,7 +22,7 @@
 
 extern "C" {
 int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size);
-int LLVMFuzzerInitialize(int* argc, char*** argv) __attribute__((weak_import));
+int LLVMFuzzerInitialize(int* argc, char*** argv) __attribute__((weak));
 }
 
 namespace {

--- a/fuzzing/repositories.bzl
+++ b/fuzzing/repositories.bzl
@@ -72,7 +72,7 @@ def rules_fuzzing_dependencies(oss_fuzz = True, honggfuzz = True, jazzer = False
         maybe(
             http_archive,
             name = "jazzer",
-            sha256 = "c8d187663365f625bb00d6b531ede17898d87bbaed59de16a85590117cff0fa4",
-            strip_prefix = "jazzer-1b11bd6515a65f02a1cbe5aaa183180faa92a55b",
-            url = "https://github.com/CodeIntelligenceTesting/jazzer/archive/1b11bd6515a65f02a1cbe5aaa183180faa92a55b.zip",
+            sha256 = "3cb2177974e81b70627f3a2bcdae5257e047d1aa23164a84a9a7bf8b7372e329",
+            strip_prefix = "jazzer-5d23f7b6bbb188fb4976f7d8b641b0e3d7655970",
+            url = "https://github.com/CodeIntelligenceTesting/jazzer/archive/5d23f7b6bbb188fb4976f7d8b641b0e3d7655970.zip",
         )

